src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1257,12 +1279,76 @@ public bool StatusCanHaveBody(int statusCode)
 
         private void ThrowResponseAlreadyStartedException(string value)
         {
-            throw new InvalidOperationException(value + " cannot be set, response had already started.");
+            throw new InvalidOperationException($"{value} cannot be set, response has already started.");
+        }
+
+        private void RejectNonBodyTransferEncodingResponse(bool appCompleted)
+        {
+            var ex = new InvalidOperationException($"Transfer-Encoding set on a {StatusCode} non-body request.");
+            if (!appCompleted)
+            {
+                // Back out of header creation surface exeception in user code
+                _requestProcessingStatus = RequestProcessingStatus.RequestStarted;
+                throw ex;
+            }
+            else
+            {
+                ReportApplicationError(ex);
+
+                // 500 Internal Server Error
+                SetErrorResponseHeaders(statusCode: 500);
+            }
+        }
+
+        private void SetErrorResponseHeaders(int statusCode)
+        {
+            Debug.Assert(!HasResponseStarted, $"{nameof(SetErrorResponseHeaders)} called after response had already started.");
+
+            StatusCode = statusCode;
+            ReasonPhrase = null;
+
+            if (FrameResponseHeaders == null)
+            {
+                InitializeHeaders();
+            }
+
+            var responseHeaders = FrameResponseHeaders;
+            responseHeaders.Reset();
+            var dateHeaderValues = DateHeaderValueManager.GetDateHeaderValues();
+
+            responseHeaders.SetRawDate(dateHeaderValues.String, dateHeaderValues.Bytes);
+            responseHeaders.SetRawContentLength("0", _bytesContentLengthZero);
+
+            if (ServerOptions.AddServerHeader)
+            {
+                responseHeaders.SetRawServer(Constants.ServerName, _bytesServer);
+            }
+        }
+
+        public void HandleNonBodyResponseWrite(int count)
+        {
+            if (Method == "HEAD")
+            {
+                // Don't write to body for HEAD requests.
+                Log.ConnectionHeadResponseBodyWrite(ConnectionId, count);
```

Comment:
Can we have a flag so this is only logged once per request/response? Maybe log it at the end of the response so we can count the size of all the writes.